homeassistant/components/surepetcare/binary_sensor.py

```diff
@@ -0,0 +1,188 @@
+
+    @property
+    def should_poll(self):
+        """Return true."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._state
+
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return DEFAULT_DEVICE_CLASS if not self._device_class else self._device_class
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return f"{self._household_id}-{self._id}"
+
+    async def async_update(self):
+        """Get the latest data and update the state."""
+        self._state = self.hass.data[DATA_SURE_PETCARE][self._sure_type][self._id]
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+
+        @callback
+        def update():
+            """Update the state."""
+            self.async_schedule_update_ha_state(True)
+
+        # pylint: disable=attribute-defined-outside-init
+        self._async_unsub_dispatcher_connect = async_dispatcher_connect(
+            self.hass, TOPIC_UPDATE, update
+        )
+
+    async def async_will_remove_from_hass(self):
+        """Disconnect dispatcher listener when removed."""
+        # pylint: disable=using-constant-test
+        if self._async_unsub_dispatcher_connect:
+            self._async_unsub_dispatcher_connect()
+
+
+class Flap(SurePetcareBinarySensor):
+    """Sure Petcare Flap."""
+
+    def __init__(self, _id: int, name: str, household_id: int):
+        """Initialize a Sure Petcare Flap."""
+        super().__init__(
+            _id,
+            f"Flap {name.capitalize()}",
+            household_id,
+            DEVICE_CLASS_LOCK,
+            SureThingID.FLAP.name,
+        )
+
+    @property
+    def is_on(self):
+        """Return true if light is on."""
+        try:
+            return bool(self._state["locking"]["mode"] == SureLockStateID.UNLOCKED)
+        except (KeyError, TypeError):
+            return None
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        try:
+            attributes = {
+                "battery_voltage": self._state["battery"] / 4,
+                "locking_mode": self._state["locking"]["mode"],
+                "device_rssi": self._state["signal"]["device_rssi"],
+                "hub_rssi": self._state["signal"]["hub_rssi"],
+                "mac_address": self._state["mac_address"],
+                "version": self._state["version"],
+            }
+
+        except (KeyError, TypeError) as error:
+            _LOGGER.error(
+                "Error getting device state attributes from %s: %s\n\n%s",
+                self._name,
+                error,
+                self._state,
+            )
+            attributes = {"error": self._state}
```

Comment:
We usually don't expose errors that aren't a normal part of the entity life cycle in the state attributes. Errors should be logged so they can be reported by a user and addressed by a developer.